### PR TITLE
feat: add GitLab to PR generator for applicationset

### LINKED ISSUE #9262: ApplicationSet Pull Request Generator for GitLab
# Summary

There is a mention of the possibility to use GitLab as a source for the Pull Request Generator in the [docs](https://github.com/argoproj/argo-cd/tree/master/applicationset/services/pull_request), still, there is no code for this type of generator present in the [corresponding folder](https://github.com/argoproj/argo-cd/tree/master/applicationset/services/pull_request). Accordingly, it is currently not possible to use GitLab as a source.

# Motivation

This would make it possible for people hosting their repositories on GitLab to use the Pull Request Generator.

# Proposal

Like for GitLab, Bitbucket and GitTea there should be an option to use GitLab (SaaS and SelfHosted) as a source for the ApplicationSet Pull Request Generator.

### PR DESCRIPTION
**Pull request by iverberk**
Closes #9262 

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [X] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes: #9262 (Created by Spork)
* [X] The title of the PR states what changed and the related issues number (used for the release note).
* [X] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [X] Does this PR require documentation updates?
* [X] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [X] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [X] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [X] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by RuBiCK** [comment]
Is there any ETA for this feature?

File: applicationset/services/pull_request/gitlab_test.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+	mux.HandleFunc(path, func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, path+"?labels=feature%2Cready&per_page=100", r.URL.RequestURI())
+		writeMRListResponse(t, w)
+	})
+
+	svc, err := NewGitLabService(context.Background(), "", server.URL, "278964", []string{"feature", "ready"})
+	assert.NoError(t, err)
+
+	_, err = svc.List(context.Background())
```
Comment (crenshaw-dev) [inline applicationset/services/pull_request/gitlab_test.go:101]: Can you add an assertion to show that the label filtering worked?

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Just one small nit.  :-)

**Comment by crenshaw-dev** [comment]
@RuBiCK by default, this would end up in 2.5. If you need it in 2.4, I can check with folks about cherry-picking it back.

**Comment by RuBiCK** [comment]
@crenshaw-dev that is what we need to create ephemeral environments. 
If it could be implemented in 2.4 it will be really amazing.

File: applicationset/services/pull_request/gitlab_test.go (hunk 1 above)
Comment (crenshaw-dev) [reply applicationset/services/pull_request/gitlab_test.go:101]: Ah my bad for reading too quickly. Feel free to mock that behavior, but I don't think it's necessary. Thanks for the quick response!

**Comment by RuBiCK** [comment]
@crenshaw-dev thanks for bringing that up in the meeting, I really appreciate it.

In the beginning, I thought it was an undocumented feature because it's mentioned [in the documentation](https://github.com/argoproj/argo-cd/blob/87d771017581162718d936351bc9f9c96f8e3ea8/docs/operator-manual/applicationset/Generators-Pull-Request.md) and I thought It would be already included and documented for 2.4.

In my company, we do not program in GO and I'm not a developer, but I'll try to find someone to help me build this version because It's crucial for the company this feature to adopt Argocd.

**Comment by crenshaw-dev** [comment]
> In the beginning, I thought it was an undocumented feature because it's mentioned [in the documentation](https://github.com/argoproj/argo-cd/blob/87d771017581162718d936351bc9f9c96f8e3ea8/docs/operator-manual/applicationset/Generators-Pull-Request.md) and I thought It would be already included and documented for 2.4.

Ah yep, we need to remove that. I'm not sure how it got there.

The basic steps of building this internally are:
1) clone the argo-cd repo and checkout the branch for the release you want to run (release-2.4 probably)
2) `git cherry-pick COMMIT` onto that branch, where `COMMIT` is the commit has for this PR (that commit hash will appear on this PR after I merge it)
3) `make image IMAGE_TAG=2.4-plus-gitlab`
4) `docker push` that tag to your internal artifact repo
5) update your Argo CD install manifests to use that image for the appset controller (no need to update the other images)
6) `kubectl apply` the applicationset CRD manifest from this PR to the cluster where you need gitlab support

**Review by crenshaw-dev — APPROVED** [review]
Thanks!

**Pull request merged**